diff --git a/imgui.cpp b/imgui.cpp
index 52be84ba8..3b440247b 100644
--- a/imgui.cpp
+++ b/imgui.cpp
@@ -15466,6 +15466,7 @@ static void ImGui::DockContextBuildNodesFromSettings(ImGuiContext* ctx, ImGuiDoc
         else if (node->ParentNode && node->ParentNode->ChildNodes[1] == NULL)
             node->ParentNode->ChildNodes[1] = node;
         node->SelectedTabId = settings->SelectedTabId;
+        node->TabIdFromSettings = settings->SelectedTabId;
         node->SplitAxis = (ImGuiAxis)settings->SplitAxis;
         node->SetLocalFlags(settings->Flags & ImGuiDockNodeFlags_SavedFlagsMask_);
 
@@ -16553,7 +16554,16 @@ static void ImGui::DockNodeUpdate(ImGuiDockNode* node)
         node->IsFocused = false;
     }
     if (node->TabBar && node->TabBar->SelectedTabId)
-        node->SelectedTabId = node->TabBar->SelectedTabId;
+    {
+        // restore selected tab from previous session
+        if (node->TabIdFromSettings) {
+            node->TabBar->SelectedTabId = node->SelectedTabId = node->TabIdFromSettings;
+            node->TabIdFromSettings = 0;
+        }
+        else {
+            node->SelectedTabId = node->TabBar->SelectedTabId;
+        }
+    }
     else if (node->Windows.Size > 0)
         node->SelectedTabId = node->Windows[0]->TabId;
 
diff --git a/imgui_internal.h b/imgui_internal.h
index 3190224ff..f5088fd5d 100644
--- a/imgui_internal.h
+++ b/imgui_internal.h
@@ -1685,6 +1685,7 @@ struct IMGUI_API ImGuiDockNode
     int                     LastFrameFocused;           // Last frame number the node was focused.
     ImGuiID                 LastFocusedNodeId;          // [Root node only] Which of our child docking node (any ancestor in the hierarchy) was last focused.
     ImGuiID                 SelectedTabId;              // [Leaf node only] Which of our tab/window is selected.
+    ImGuiID                 TabIdFromSettings;
     ImGuiID                 WantCloseTabId;             // [Leaf node only] Set when closing a specific tab/window.
     ImGuiID                 RefViewportId;              // Reference viewport ID from visible window when HostWindow == NULL.
     ImGuiDataAuthority      AuthorityForPos         :3;
